pkg/websocket/client.go

```diff
@@ -144,35 +144,49 @@ func (c *Client) Run(ctx context.Context) {
 			close(c.send)
 			close(c.stopReadPump)
 			close(c.stopWritePump)
-
+			c.SendCloseFrame(ws.CloseNormalClosure, "Connection Done")
 			return
 		case <-c.done:
 			close(c.send)
 			close(c.stopReadPump)
 			close(c.stopWritePump)
 			close(c.NotifyExpired)
-
+			c.SendCloseFrame(ws.CloseNormalClosure, "Connection Done")
 			return
 		case <-c.notifyClose:
 			c.cfg.Log.WithFields(log.Fields{
 				"prefix": "websocket.client.Run",
 			}).Debug("Disconnected from Stripe")
 			close(c.stopReadPump)
 			close(c.stopWritePump)
+			c.SendCloseFrame(ws.CloseGoingAway, "Server closed the connection")
 			c.wg.Wait()
 		case <-time.After(c.cfg.ReconnectInterval):
 			c.cfg.Log.WithFields(log.Fields{
 				"prefix": "websocket.Client.Run",
 			}).Debug("Resetting the connection")
 			close(c.stopReadPump)
 			close(c.stopWritePump)
+			c.SendCloseFrame(ws.CloseNormalClosure, "Resetting the connection")
+			c.wg.Wait()
+		}
+	}
+}
 
-			if c.conn != nil {
-				c.conn.Close() // #nosec G104
-			}
+// SendCloseFrame executes a proper closure handshake then closes the connection
+// list of close codes: https://datatracker.ietf.org/doc/html/rfc6455#section-7.4
+func (c *Client) SendCloseFrame(closeCode int, text string) {
+	if c.conn != nil {
+		message := ws.FormatCloseMessage(closeCode, text)
 
-			c.wg.Wait()
+		err := c.conn.WriteControl(ws.CloseMessage, message, time.Now().Add(time.Second))
+		if err != nil {
+			c.cfg.Log.WithFields(log.Fields{
+				"prefix": "websocket.Client.Run",
+				"error":  err,
+			}).Debug("Error while trying to send close frame")
 		}
+		c.conn.Close()
```

Comment:
is it safe/reasonable to send the close frame and close the connection at approximately the same time? just looking at https://github.com/gorilla/websocket/blob/master/examples/echo/client.go it doesn't seem to do that (though I can imagine wanting a deadline not to be exceeded)

also for my own curiosity what's the difference between WriteControl and WriteMessage?

Reply:
Looking into the safety question. 

Re: the diff between WriteControl and WriteMessage, I initially used WriteControl because the docs say that a close is a type of control message: https://github.com/gorilla/websocket/blob/master/conn.go#L73

I think the only difference is that WriteControl is concurrency-safe: https://pkg.go.dev/github.com/gorilla/websocket#hdr-Control_Messages

Reply:
I think you're right, this other example has a wait period too: https://github.com/gorilla/websocket/blob/5ade364dff1f03abefdae800ed19074bf707b08f/examples/command/main.go#L75-L78

https://github.com/gorilla/websocket/blob/5ade364dff1f03abefdae800ed19074bf707b08f/examples/command/main.go#L38-L39

Reply:
ok that makes sense, thanks for the clarifications

Reply:
Thanks so much for pointing that out btw! https://github.com/gorilla/websocket/issues/624#issuecomment-675728014